qiskit_experiments/visualization/plotters/base_plotter.py

```diff
@@ -0,0 +1,453 @@
+
+    def data_exists_for(self, series_name: str, data_keys: Union[str, List[str]]) -> bool:
+        """Returns whether the given data-keys exist for the given series.
+
+        Args:
+            series_name: The name of the given series.
+            data_keys: The data-keys to be checked.
+
+        Returns:
+            bool: True if all data-keys have values assigned for the given series. False if at least one
+                does not have a value assigned.
+        """
+        if not isinstance(data_keys, list):
+            data_keys = [data_keys]
+
+        # Handle non-existent series name
+        if series_name not in self._series_data:
+            return False
+
+        return all(key in self._series_data[series_name] for key in data_keys)
+
+    @abstractmethod
+    def _plot_figure(self):
+        """Generates a figure using :attr:`drawer` and :meth:`data`.
+
+        Sub-classes must override this function to plot data using the drawer. This function is called by
+        :meth:`figure` when :attr:`drawer` can be used to draw on the canvas.
+        """
+
+    def figure(self) -> Any:
+        """Generates and returns a figure for the already provided series and figure data.
+
+        :meth:`figure` calls :meth:`_plot_figure`, which is overridden by sub-classes. Before and after
+        calling :meth:`_plot_figure`; :func:`_initialize_drawer`, :func:`initialize_canvas` and
+        :func:`format_canvas` are called on the drawer respectively.
+
+        Returns:
+            Any: A figure generated by :attr:`drawer`, of the same type as ``drawer.figure``.
+        """
+        # Initialize drawer, to copy axis, subplots, and plot-options across.
+        self._initialize_drawer()
+
+        # Initialize canvas, which creates subplots, assigns axis labels, etc.
+        self.drawer.initialize_canvas()
+
+        # Plot figure for given subclass. This is the core of BasePlotter subclasses.
+        self._plot_figure()
+
+        # Final formatting of canvas, which sets axis limits etc.
+        self.drawer.format_canvas()
+
+        # Return whatever figure is created by the drawer.
+        return self.drawer.figure
+
+    @classmethod
+    @abstractmethod
+    def expected_series_data_keys(cls) -> List[str]:
+        """Returns the expected series data-keys supported by this plotter."""
+
+    @classmethod
+    @abstractmethod
+    def expected_figure_data_keys(cls) -> List[str]:
+        """Returns the expected figures data-keys supported by this plotter."""
+
+    @property
+    def options(self) -> Options:
+        """Options for the plotter.
+
+        Options for a plotter modify how the class generates a figure. This includes an optional axis
+        object, being the drawer canvas. Make sure verify whether the option you want to set is in
+        :attr:`options` or :attr:`plot_options`.
+        """
+        return self._options
+
+    @property
+    def plot_options(self) -> Options:
```

Comment:
Give the description, should we rename this to `drawer_options` or `options_for_drawers`? Same for the associated internal variable?

Reply:
@nkanazawa1989 also brought this up in another comment. I've renamed `plot_options` to `figure_options` based on his suggestion. Looking back at the code, it would be good if users of `BasePlotter` were oblivious to drawers, unless otherwise necessary. See 956696a for the changes.